pkg/eventfilter/subscriptionsapi/cesql_correlationid_filter.go

```diff
@@ -0,0 +1,227 @@
+
+	var correlationIdFilterFunction = cefn.NewFunction(
+		"KN_VERIFY_CORRELATIONID",
+		[]cesql.Type{cesql.StringType},
+		cesql.TypePtr(cesql.StringType),
+		cesql.BooleanType,
+		func(event cloudevents.Event, i []interface{}) (interface{}, error) {
+			correlationId := i[0].(string)
+
+			match, _ := regexp.MatchString(".*:.*", correlationId)
+			if !match {
+				return false, errors.New("correlationId Format: <original id>:<base64/hex encoded encrypted id>")
+			}
+
+			slice := strings.Split(correlationId, ":")
+			originalId := slice[0]
+			encryptedId := slice[1]
+
+			encryptedIdBytes, err := decodeBase64OrHex(encryptedId)
+			if err != nil {
+				return false, err
+			}
+
+			// Create a set of secret names to try looking for in k8s
+			secretNamesToTry := make(map[string]bool)
+
+			// Iterate through secret names in argument list and add them to the set
+			for num, secret := range i {
+				if num != 0 {
+					secretNamesToTry[secret.(string)] = true
+				}
+			}
+
+			secrets, err := getSecretsFromK8s()
+			if err != nil {
+				return false, err
+			}
+
+			/*
+			 * Go through each retrived secret
+			 * Check if the secret has data for a key and algorithm
+			 * Check if the secret name matches with one of the arguments
+			 * Decrypt encryptedId using the current secret's key
+			 * Check if the encrypted value matches with originalId
+			 */
+			for _, secret := range secrets {
+				key, keyFieldExists := secret.Data["key"]
+				algorithm, algorithmFieldExists := secret.Data["algorithm"]
+
+				if keyFieldExists && algorithmFieldExists && secretNamesToTry[secret.Name] {
+					var decryptionFunc func(originalId string, encryptedIdBytes []byte, key []byte) (bool, error)
+					switch strings.ToUpper(string(algorithm)) {
+					case "AES", "AES-ECB":
+						decryptionFunc = compareWithAES
+					case "DES":
+						decryptionFunc = compareWithDES
+					case "3DES", "TRIPLEDES":
+						decryptionFunc = compareWithTripleDES
+					case "RC4":
+						decryptionFunc = compareWithRC4
+					default:
+						return false, errors.New("cipher algorithm not supported")
+					}
+
+					res, err := decryptionFunc(originalId, encryptedIdBytes, key)
+					if err != nil {
+						return false, err
+					}
+					if res {
+						return true, nil
+					}
+				}
+			}
+
+			return false, nil
+		},
+	)
+
+	ceruntime.AddFunction(correlationIdFilterFunction)
+
+	return NewCESQLFilter(expr)
+}
+
+func getSecretsFromK8s() ([]v1.Secret, error) {
+	// Assuming .kube/config is in the home directory
+	basePath, _ := os.UserHomeDir()
+	defaultKubeConfigPath := path.Join(basePath, ".kube", "config")
+
+	// Set up k8s client to get secrets
+	config, err := clientcmd.BuildConfigFromFlags("", defaultKubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	secrets, err := clientset.CoreV1().Secrets("default").List(context.Background(), metav1.ListOptions{})
```

Comment:
Probably need to configure:
1. which namespace to look up secrets in
2. which secret names to get